# Business Analytics - Assignment 2  

**Assignment Points**: 100  
**Submission**: Provide your answers in this notebook and submit it via iLearn

- Where a question requires a written (text) solution provide your answer in Markdown in appropriate cells under each question.
- Comment out your print statements unless you are explicitly asked to use the print() function. 
- 5 marks will be deducted for printed outputs that are not asked for.

### About the Assignment

- Assignment 2 extends Assignment 1 on credit card applications. 


- For this assignment there are two files in the `data` folder `credit_record.csv` and `application_record.csv` where bank clients are related by the `ID` column.
- In `credit_record.csv` we have the following variables

| Feature Name         | Explanation     | Additional Remarks |
|--------------|-----------|-----------|
| ID | Randomly allocated client number      |         |
| AMT_INCOME_TOTAL   | Annual income  |  |
| NAME_INCOME_TYPE   | Income Source |  |
| NAME_EDUCATION_TYPE   | Level of Education  |  |
| CODE_GENDER   | Applicant's Gender   |  |
| FLAG_OWN_CAR | Car Ownership |  | 
| CNT_CHILDREN | Number of Children | |
| FLAG_OWN_REALTY | Real Estate Ownership | | 
| NAME_FAMILY_STATUS | Relationship Status | | 
| NAME_HOUSING_TYPE | Housing Type | | 
| DAYS_BIRTH | No. of Days | Count backwards from current day (0), -1 means yesterday
| DAYS_EMPLOYED | No. of Days | Count backwards from current day(0). If positive, it means the person is currently unemployed.
| FLAG_MOBIL | Mobile Phone Ownership | | 
| FLAG_WORK_PHONE | Work Phone Ownership | | 
| FLAG_PHONE | Landline Phone Ownership | | 
| FLAG_EMAIL | Landline Phone Ownership | | 
| OCCUPATION_TYPE | Occupation | | 
| CNT_FAM_MEMBERS | Count of Family Members | |



- In `credit_record.csv` we have the following variables


| Feature Name         | Explanation     | Additional Remarks |
|--------------|-----------|-----------|
| ID | Randomly allocated client number | |
| MONTHS_BALANCE | Number of months in the past from now when STATUS is measured | 0 = current month, -1 = last month, -2 = two months ago, etc.|
| STATUS | Number of days a payment is past due | 0: 1-29 days past due 1: 30-59 days past due 2: 60-89 days overdue 3: 90-119 days overdue 4: 120-149 days overdue 5: Overdue or bad debts, write-offs for more than 150 days C: paid off that month X: No loan for the month |

---
---

### Task 1: Reading, Summarising and Cleaning Data (Total Marks: 30)



**Question 1.** 

1. Import the `application_record.csv` and `credit_record.csv` files from `data` folder into pandas DataFrames named `df_application` and `df_credit`, respectively. (1 mark)

2. How many rows are there in `df_application` and `df_credit`, respectively? Answer using both print() function and in Markdown text. (1 mark)

3. How many unique bank clients are there in `df_application` and `df_credit`? Answer using both print() function and in Markdown text. (1 mark)

4. Add the records from `df_credit` to `df_application` by merging the data from the two DataFrames on the `ID` column, and output the joint data into a new DataFrame named `df`. Hint: Use `merge` function from pandas by setting `how` parameter to `inner` (4 marks) 

5. How many rows and how many unique clients are there in `df`? (1 mark)

6. How are multiple rows for each `ID` in `df` different? Answer in Markdown text. (2 mark) 

(10 marks)


In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.decomposition import PCA
from sklearn.metrics import accuracy_score

/opt/anaconda3/lib/python3.9/site-packages/pandas/core/computation/expressions.py:21: UserWarning: Pandas requires version '2.8.4' or newer of 'numexpr' (version '2.8.3' currently installed).
  from pandas.core.computation.check import NUMEXPR_INSTALLED
/opt/anaconda3/lib/python3.9/site-packages/pandas/core/arrays/masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).
  from pandas.core import (
/opt/anaconda3/lib/python3.9/site-packages/scipy/__init__.py:155: UserWarning: A NumPy version >=1.18.5 and <1.25.0 is required for this version of SciPy (detected version 1.26.4
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


**1.1**

In [2]:
df_application= pd.read_csv('application_record.csv')
df_credit= pd.read_csv('credit_record.csv')

**1.2**

In [3]:
print(f"Number of rows in df_application: {len(df_application)}")
print(f"Number of rows in df_credit: {len(df_credit)}")

Number of rows in df_application: 438557
Number of rows in df_credit: 1048575


There are 438,557 rows in df_application and 1,048,575 rows in df_credit. 

**1.3**

In [4]:
df_application['ID'].nunique()

438510

In [5]:
df_credit['ID'].nunique()

45985

There are 438510 unique bank clients in df_application and 45985 unique bank clients in df_credit

**1.4**

In [6]:
df = pd.merge(df_application, df_credit, on='ID', how='inner')
df

ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
0       5008804           M            Y               Y             0   
1       5008804           M            Y               Y             0   
2       5008804           M            Y               Y             0   
3       5008804           M            Y               Y             0   
4       5008804           M            Y               Y             0   
...         ...         ...          ...             ...           ...   
777710  5150337           M            N               Y             0   
777711  5150337           M            N               Y             0   
777712  5150337           M            N               Y             0   
777713  5150337           M            N               Y             0   
777714  5150337           M            N               Y             0   

        AMT_INCOME_TOTAL NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
0               427500.0          Working               Higher education   
1               427500.0          Working               Higher education   
2               427500.0          Working               Higher education   
3               427500.0          Working               Higher education   
4               427500.0          Working               Higher education   
...                  ...              ...                            ...   
777710          112500.0          Working  Secondary / secondary special   
777711          112500.0          Working  Secondary / secondary special   
777712          112500.0          Working  Secondary / secondary special   
777713          112500.0          Working  Secondary / secondary special   
777714          112500.0          Working  Secondary / secondary special   

          NAME_FAMILY_STATUS NAME_HOUSING_TYPE  DAYS_BIRTH  DAYS_EMPLOYED  \
0             Civil marriage  Rented apartment      -12005          -4542   
1             Civil marriage  Rented apartment      -12005          -4542   
2             Civil marriage  Rented apartment      -12005          -4542   
3             Civil marriage  Rented apartment      -12005          -4542   
4             Civil marriage  Rented apartment      -12005          -4542   
...                      ...               ...         ...            ...   
777710  Single / not married  Rented apartment       -9188          -1193   
777711  Single / not married  Rented apartment       -9188          -1193   
777712  Single / not married  Rented apartment       -9188          -1193   
777713  Single / not married  Rented apartment       -9188          -1193   
777714  Single / not married  Rented apartment       -9188          -1193   

        FLAG_MOBIL  FLAG_WORK_PHONE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  \
0                1                1           0           0             NaN   
1                1                1           0           0             NaN   
2                1                1           0           0             NaN   
3                1                1           0           0             NaN   
4                1                1           0           0             NaN   
...            ...              ...         ...         ...             ...   
777710           1                0           0           0        Laborers   
777711           1                0           0           0        Laborers   
777712           1                0           0           0        Laborers   
777713           1                0           0           0        Laborers   
777714           1                0           0           0        Laborers   

        CNT_FAM_MEMBERS  MONTHS_BALANCE STATUS  
0                     2               0      C  
1                     2              -1      C  
2                     2              -2      C  
3                     2              -3      C  
4                     2              -4      C  
...                 ...             ...    

**1.5**

In [7]:
df['ID'].nunique()

36457

There are 777715 rows and 36457 unique clients in df. 

**1.6**

**How are multiple rows for each ID in df different?**

In the merged DataFrame (`df`), multiple rows for each ID are different because each ID can have multiple credit records associated with it. Each row in the `df_credit` DataFrame represents a different credit record for a client, while the `df_application` DataFrame contains the application information for each client. When merged, the resulting DataFrame will have multiple rows for each client ID, where each row combines the client's application information with one of their credit records.

---

**Question 2.**

1. Change the values of `STATUS` in `df` according to the following mapping: {C, X, 0} -> 0 and {1, 2, 3, 4, 5} -> 1 making sure that the new values of 0 and 1 are encoded as integers. (2 marks)
2. Create a new numpy array named `list_of_defaults` containing *unique* `ID` numbers for the clients who have `STATUS` = 1 in any of the last 12 months in the dataset. (2 marks) 
3. Create a new DataFrame called `df_final` that contains the rows of `df` for which the `ID` are in `list_of_defaults`, keeping only one row for each `ID` (i.e. eliminate rows with duplicate `ID`s while keeping the first duplicate row). How many rows do you have in `df_final`? Answer using both print() function and in Markdown text. (Hint: find out about `isin()` function in pandas.) (2 marks)
4. Add a new column `y = 1` for all the rows in `df_final`. (1 marks)
5. Increase `df_final` to a total of 4,000 rows by adding rows from `df` with unique `ID`s (nonduplicated `ID`s) which are not in `list_of_defaults`. To do this start adding the rows from the beginning of `df`. (Hint: learn what `~`, i.e. tilde sign, does in pandas). (2 marks) 
6. Fill the missing values of `y` in `df_final` with zeros. Remove `STATUS` and `MONTHS_BALANCE` from `df_final`. How many clients with  overdue payments of more than 29 days and how many clients with less than 29 days overdue payments are there in `df_final`? Answer using both print() function and in Markdown text.(1 mark)

(10 marks)

**2.1**

In [8]:
df['STATUS'] = df['STATUS'].map({'C': 0, 'X': 0, '0': 0, '1': 1, '2': 1, '3': 1, '4': 1, '5': 1}).astype(int)

**2.2**

In [9]:
last_12_months = df['MONTHS_BALANCE'].max() - 12
list_of_defaults = df[(df['MONTHS_BALANCE'] >= last_12_months) & (df['STATUS'] == 1)]['ID'].unique()

**2.3**

In [10]:
df_final = df[df['ID'].isin(list_of_defaults)].drop_duplicates(subset='ID', keep='first')

In [11]:
print(f"Number of rows in df_final: {len(df_final)}")

Number of rows in df_final: 1833


Number of rows in df_final: 1833

**2.4**

In [12]:
df_final['y'] = 1

**2.5**

In [13]:
additional_rows = df[~df['ID'].isin(list_of_defaults)].drop_duplicates(subset='ID', keep='first').head(4000 - len(df_final))
df_final = pd.concat([df_final, additional_rows])

**2.6**

In [14]:
df_final['y'].fillna(0, inplace=True)

/var/folders/8t/glsx5ls50qqbbbhkvwpkhdnc0000gn/T/ipykernel_3304/1979685076.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_final['y'].fillna(0, inplace=True)


In [15]:
df_final.drop(columns=['STATUS', 'MONTHS_BALANCE'], inplace=True)

In [16]:
overdue_29_days = df_final['y'].sum()
not_overdue_29_days = len(df_final) - overdue_29_days
print(f"Clients with overdue payments of more than 29 days: {overdue_29_days}")
print(f"Clients with less than 29 days overdue payments: {not_overdue_29_days}")

Clients with overdue payments of more than 29 days: 1833.0
Clients with less than 29 days overdue payments: 2167.0


There are 1,833 people with overdue payments of more than 29 days and 2167 people with overdue payments of less than 29 days

<hr style="width:25%;margin-left:0;"> 

**Question 3**. 
- Delete `ID` column from `df_final` (1 marks)
- Of the remaining variables in `df_final` and assuming that `NAME_EDUCATION_TYPE` is the only ordinal variable, how many variable are of numeric and nominal types? Provide lists of all numeric and nominal variables. (6)
- Using an appropriate function find and comment on the missing values in `df_final`, i.e. how many variables and how many observations? (3 marks)   
(10 marks)

**3.1**

In [17]:
df_final.drop(columns=['ID'], inplace=True)

**3.2**

In [18]:
numeric_vars = df_final.select_dtypes(include=[np.number]).columns.tolist()
nominal_vars = df_final.select_dtypes(include=[object]).columns.tolist()
nominal_vars.remove('NAME_EDUCATION_TYPE')

print(f"Numeric variables: {numeric_vars}")
print(f"Nominal variables: {nominal_vars}")

Numeric variables: ['CNT_CHILDREN', 'AMT_INCOME_TOTAL', 'DAYS_BIRTH', 'DAYS_EMPLOYED', 'FLAG_MOBIL', 'FLAG_WORK_PHONE', 'FLAG_PHONE', 'FLAG_EMAIL', 'CNT_FAM_MEMBERS', 'y']
Nominal variables: ['CODE_GENDER', 'FLAG_OWN_CAR', 'FLAG_OWN_REALTY', 'NAME_INCOME_TYPE', 'NAME_FAMILY_STATUS', 'NAME_HOUSING_TYPE', 'OCCUPATION_TYPE']


There are 11 variables of numeric type and 7 of nominal type.

**3.3**

In [19]:
missing_values = df_final.isnull().sum()
num_missing_vars = missing_values[missing_values > 0].count()
num_missing_observations = missing_values[missing_values > 0].sum()

print(f"Number of variables with missing values: {num_missing_vars}")
print(f"Number of missing observations: {num_missing_observations}")

Number of variables with missing values: 1
Number of missing observations: 1167


The number of variables with missing values is 1 and the number of missing onservations is 1,167. 



---
---

### Task 2: Imputing missing values and dealing with categorical features (Total Marks: 30)



**Question 4.** 
- Use an appropriate `pandas` function to impute missing values in `df_final` (10 marks)
    - Be careful when deciding which method to use to replace missing observations 
    - Take into consideration the type of each variable and the best practices we discussed in class/lecture notes
- Briefly explain what you have done and why. (5 marks)

(Total: 15 marks)

**4.1**

In [20]:
impute_values = {var: df_final[var].median() for var in numeric_vars}
impute_values.update({var: df_final[var].mode()[0] for var in nominal_vars})
df_final.fillna(value=impute_values, inplace=True)

In [21]:
missing_values_after = df_final.isnull().sum()
num_missing_vars_after = missing_values_after[missing_values_after > 0].count()
num_missing_observations_after = missing_values_after[missing_values_after > 0].sum()

print(f"Number of variables with missing values after imputation: {num_missing_vars_after}")
print(f"Number of missing observations after imputation: {num_missing_observations_after}")

Number of variables with missing values after imputation: 0
Number of missing observations after imputation: 0


**4.2**

The selection of methods is based on best practices for handling missing data. These methods maintain the integrity and distribution of the dataset, ensuring the imputed values are representative of the original data. By applying these methods, the impact of missing data is minimised on the analysis and maintains the overall strucutre.

**Numeric Variables**

Median is used for numeric variables to impute missing values. The median is used particularly when dealing with skewed data or outliers. This method ensures the imputed values do not overly influence the distribution of the data. 

**Nominal Variables**

For nominal variables, the method mode was used as it was the most appropriate to impute for categorial data, preserving the distribution of the categories. 

<hr style="width:25%;margin-left:0;"> 

**Question 5**. Convert the values in `NAME_EDUCATION_TYPE` as follows
- Lower secondary -> 1
- Secondary / secondary special -> 2
- Incomplete higher -> 3
- Higher education -> 4


(Total: 5 marks)  

In [22]:
education_type_mapping = {
    'Lower secondary': 1,
    'Secondary / secondary special': 2,
    'Incomplete higher': 3,
    'Higher education': 4}

In [23]:
df_final['NAME_EDUCATION_TYPE'] = df_final['NAME_EDUCATION_TYPE'].replace(education_type_mapping)

<hr style="width:25%;margin-left:0;"> 

**Question 6**. 

Add dummy variables to `df_final` for all of the nominal features which are currently stored as string (text). 
- Make sure to delete the original variables from the dataframe
- Drop the first column from each set of created dummy variable, i.e. for each feature



(Total: 10 marks)  

In [24]:
nominal_vars = df_final.select_dtypes(include=[object]).columns.tolist()
df_final = pd.get_dummies(df_final, columns=nominal_vars, drop_first=True)

---
---

### Task 3 Preparing X and y arrays (Total Marks: 10)

**Question 7**. 

- Create a numpy array named `y` from the `y` column of `df_final` making sure that the values of the array `y` are stored as integers (3 marks)   
- Create a numpy array named `X`  from all the remaining features in `df_final` (2 marks)   

(Total: 5 Marks)

In [25]:
y = df_final['y'].astype(int).values
X = df_final.drop(columns=['y']).values

print(f"Shape of y: {y.shape}")
print(f"Shape of X: {X.shape}")

Shape of y: (4000,)
Shape of X: (4000, 46)


<hr style="width:25%;margin-left:0;"> 

**Question 8**. 

- Use an appropriate scikit-learn library we used in class to create `y_train`, `y_test`, `X_train` and `X_test` by splitting the data into 70% train and 30% test datasets (2.5 marks) 
    - Set random_state to 7 and stratify the subsamples so that train and test datasets have roughly equal proportions of the target's class labels 
- Standardise the data using `StandardScaler` library (2.5 marks)   

(Total: 5 marks) 

In [26]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=7, stratify=y)

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

print(f"Shape of X_train: {X_train.shape}")
print(f"Shape of X_test: {X_test.shape}")
print(f"Shape of y_train: {y_train.shape}")
print(f"Shape of y_test: {y_test.shape}")

Shape of X_train: (2800, 46)
Shape of X_test: (1200, 46)
Shape of y_train: (2800,)
Shape of y_test: (1200,)


---
---

### Task 4. Support Vector Classifier and Accuracies (Total Marks: 30)


**Question 9**. 

- Train a Support Vector Classifier on standardised data (3 marks)
    - Use `linear` kernel and set `random_state` to 7 (don't change any other parameters)
    - Compute and print training and test dataset accuracies
- Train another Support Vector Classifier on standardised data (3 marks)
    - Use `rbf` kernel and set `random_state` to 7 (don't change any other parameters)
    - Compute and print training and test dataset accuracies
- What can you say about the presence of nonlinearities in the dataset? (4 marks)

(Total: 10 marks)  

**9.1**

In [27]:
svc_linear = SVC(kernel='linear', random_state=7)
svc_linear.fit(X_train, y_train)

train_accuracy_linear = accuracy_score(y_train, svc_linear.predict(X_train))
test_accuracy_linear = accuracy_score(y_test, svc_linear.predict(X_test))

print(f"Linear Kernel - Training Accuracy: {train_accuracy_linear}")
print(f"Linear Kernel - Test Accuracy: {test_accuracy_linear}")

Linear Kernel - Training Accuracy: 0.5867857142857142
Linear Kernel - Test Accuracy: 0.5658333333333333


**9.2**

In [28]:
svc_rbf = SVC(kernel='rbf', random_state=7)
svc_rbf.fit(X_train, y_train)

train_accuracy_rbf = accuracy_score(y_train, svc_rbf.predict(X_train))
test_accuracy_rbf = accuracy_score(y_test, svc_rbf.predict(X_test))

print(f"RBF Kernel - Training Accuracy: {train_accuracy_rbf}")
print(f"RBF Kernel - Test Accuracy: {test_accuracy_rbf}")

RBF Kernel - Training Accuracy: 0.7760714285714285
RBF Kernel - Test Accuracy: 0.6958333333333333


**9.3**

### Analysis Output

#### RBF Kernel Performance
- **Training Accuracy:** 77.6%
- **Test Accuracy:** 69.58%

#### Linear Kernel Performance
- **Training Accuracy:** 58.68%
- **Test Accuracy:** 56.58%

The RBF's better performance on both the training and test sets suggests that the model can generalise well, capturing the essential patterns in the data without overfitting excessively.  The higher accuracy with the RBF kernel performance indicates the presence of nonlinear relationships within the dataset. These nonlinearities are effectively captured by the RBF kernel, leading to better performance. 

The linear kernel's poor performance highlights its limitations in capturing complex patterns. The dataset likely includes interactions and relationships between features that are inherently nonlinear, which the linear kernel fails to model. 

#### Conclusion
The performance gap between the RBF kernel and the linear kernel suggests the presence of significant nonlinearities in the dataset. The RBF kernel's ability to capture these nonlinear relationships leads to higher accuracies, demonstrating the importance of using more flexible model to adaquately represent the underlying structure of the data. 

<hr style="width:25%;margin-left:0;"> 

**Question 10**

- Extract 2 linear principal components from the standardised features using an appropriate `sklearn` library (5 marks)
- Train a Support Vector Classifier on the computed principal components (5 marks) 
    - Use `rbf` kernel and set `random_state` to 7 (don't change any other parameters)
- Compute and print training and test dataset accuracies (5 marks)
- What can you say about the ability of the 2 principal components to compress the information contained in the features matrix `X`, and why? (5 marks)     


(Total: 20 marks)  

**10.1**

In [29]:
pca = PCA(n_components=2, random_state=7)
X_train_pca = pca.fit_transform(X_train)
X_test_pca = pca.transform(X_test)

**10.2**

In [30]:
svc_pca = SVC(kernel='rbf', random_state=7)
svc_pca.fit(X_train_pca, y_train)

SVC(random_state=7)

**10.3**

In [31]:
train_accuracy_pca = accuracy_score(y_train, svc_pca.predict(X_train_pca))
test_accuracy_pca = accuracy_score(y_test, svc_pca.predict(X_test_pca))

print(f"Training Accuracy on PCA components: {train_accuracy_pca}")
print(f"Test Accuracy on PCA components: {test_accuracy_pca}")

Training Accuracy on PCA components: 0.58
Test Accuracy on PCA components: 0.5733333333333334


In [32]:
explained_variance_ratio = pca.explained_variance_ratio_
print(f"Explained variance ratio of the 2 principal components: {explained_variance_ratio}")
print(f"Total explained variance by the 2 principal components: {explained_variance_ratio.sum()}")

if explained_variance_ratio.sum() < 0.90:
    print("The 2 principal components do not capture sufficient information from the features matrix X.")
else:
    print("The 2 principal components capture a significant amount of information from the features matrix X.")

Explained variance ratio of the 2 principal components: [0.08755806 0.05809968]
Total explained variance by the 2 principal components: 0.14565774481327776
The 2 principal components do not capture sufficient information from the features matrix X.


**10.4**

The variance ratio displays the proportion of the dataset's variance that each principal component captures. In this case:
- **First Principle Component:** 8.76% of the variance
- **Second Principle Component:** 5.81% of the variance
- **Total Explained Variance:** 14.57%

#### Analysis Output
The two principle components together capture only 14.57% of the total variance in the dataset. This low percentage means that a significant amount (85.43%) of the variance in the original data is not represented by these two components. 

The goal of principal component analysis (PCA) is to reduce the dimensionality of the data while retaining as much information (variance) as possible. With only 14.57% of the variance retained, the two principal components fail to adequately compress the information contained in the features matrix X. 

The accuracies of the SVM classifier (58% training accuracy and 57.33% test accuracy) using the two principal components further suggest that these components do not capture enough information to perform well in a predictive model. This inadequate compression leads to a loss of crucial information that impacts the classifier's ability to distinguish between different classes effectively. 

#### Conclusion
The ability of the two principal components to compress the information contained in the features matrix X is limited. They capture only 14.57% of the variance, indicating that a significant portion of the original data's information is lost. This loss of information is reflected in the performance of the Support Vector Classifier trained on these components. To achieve better compression and retain more meaningful information, it would be necessary to use a higher number of principal components. 

---
---